origin/cmd/cmd.go

```diff
@@ -43,97 +44,91 @@ import (
 
 	"github.com/andres-erbsen/clock"
 	"github.com/pressly/chi"
-	"github.com/spf13/cobra"
 )
 
-var (
-	peerIP             string
-	peerPort           int
-	blobServerHostName string
-	blobServerPort     int
-	configFile         string
-	zone               string
-	krakenCluster      string
-	secretsFile        string
-
-	rootCmd = &cobra.Command{
-		Short: "kraken-origin serves as dedicated seeder in kraken's p2p network.",
-		Run: func(rootCmd *cobra.Command, args []string) {
-			run()
-		},
-	}
-)
-
-func init() {
-	rootCmd.PersistentFlags().StringVarP(
-		&peerIP, "peer-ip", "", "", "ip which peer will announce itself as")
-	rootCmd.PersistentFlags().IntVarP(
-		&peerPort, "peer-port", "", 0, "port which peer will announce itself as")
-	rootCmd.PersistentFlags().StringVarP(
-		&blobServerHostName, "blobserver-hostname", "", "", "optional hostname to identify origin")
-	rootCmd.PersistentFlags().IntVarP(
-		&blobServerPort, "blobserver-port", "", 0, "port which blob server listens on")
-	rootCmd.PersistentFlags().StringVarP(
-		&configFile, "config", "", "", "configuration file path")
-	rootCmd.PersistentFlags().StringVarP(
-		&zone, "zone", "", "", "zone/datacenter name")
-	rootCmd.PersistentFlags().StringVarP(
-		&krakenCluster, "cluster", "", "", "cluster name (e.g. prod01-zone1)")
-	rootCmd.PersistentFlags().StringVarP(
-		&secretsFile, "secrets", "", "", "path to a secrets YAML file to load into configuration")
+// Flags defines origin CLI flags.
+type Flags struct {
+	PeerIP             string
+	PeerPort           int
+	BlobServerHostName string
+	BlobServerPort     int
+	ConfigFile         string
+	Zone               string
+	KrakenCluster      string
+	SecretsFile        string
 }
 
-func Execute() {
-	rootCmd.Execute()
+// ParseFlags parses origin CLI flags.
+func ParseFlags() *Flags {
+	var flags Flags
+	flag.StringVar(
+		&flags.PeerIP, "peer-ip", "", "ip which peer will announce itself as")
+	flag.IntVar(
+		&flags.PeerPort, "peer-port", 0, "port which peer will announce itself as")
+	flag.StringVar(
+		&flags.BlobServerHostName, "blobserver-hostname", "", "optional hostname to identify origin")
+	flag.IntVar(
+		&flags.BlobServerPort, "blobserver-port", 0, "port which blob server listens on")
+	flag.StringVar(
+		&flags.ConfigFile, "config", "", "configuration file path")
+	flag.StringVar(
+		&flags.Zone, "zone", "", "zone/datacenter name")
+	flag.StringVar(
+		&flags.KrakenCluster, "cluster", "", "cluster name (e.g. prod01-zone1)")
+	flag.StringVar(
+		&flags.SecretsFile, "secrets", "", "path to a secrets YAML file to load into configuration")
+	flag.Parse()
+	return &flags
 }
 
-func run() {
-	if peerPort == 0 {
+// Run runs the origin.
+func Run(flags *Flags) {
+	if flags.PeerPort == 0 {
 		panic("must specify non-zero peer port")
 	}
-	if blobServerPort == 0 {
+	if flags.BlobServerPort == 0 {
```

Comment:
Now that we have a flag struct, what do you think about having a method on it like `flags.assertValid()`?

Reply:
Yeah I'll do that.

Reply:
Actually... after writing it out, I think it reads a bit better if `Run` decides which flags are required, instead of pushing this knowledge up into `Flags` (which feels somewhat detached).